src/workerd/jsg/async-context.c++

```diff
@@ -0,0 +1,254 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+#include "async-context.h"
+#include "jsg.h"
+#include "setup.h"
+#include <v8.h>
+
+namespace workerd::jsg {
+
+namespace {
+kj::Maybe<AsyncContextFrame&> tryUnwrapFrame(v8::Isolate* isolate, v8::Local<v8::Value> handle) {
+  // Gets the held AsyncContextFrame from the opaque wrappable but does not consume it.
+  KJ_IF_MAYBE(wrappable, Wrappable::tryUnwrapOpaque(isolate, handle)) {
+    OpaqueWrappable<kj::Own<AsyncContextFrame>>* holder =
+        dynamic_cast<OpaqueWrappable<kj::Own<AsyncContextFrame>>*>(wrappable);
+    KJ_ASSERT(holder != nullptr);
+    KJ_ASSERT(!holder->movedAway);
+    return *holder->value;
+  }
+  return nullptr;
+}
+
+}  // namespace
+
+AsyncContextFrame::AsyncContextFrame(IsolateBase& isolate)
+    : isolate(isolate) {}
+
+AsyncContextFrame::AsyncContextFrame(
+    Lock& js,
+    kj::Maybe<AsyncContextFrame&> maybeParent,
+    kj::Maybe<StorageEntry> maybeStorageEntry)
+    : AsyncContextFrame(IsolateBase::from(js.v8Isolate)) {
+  KJ_IF_MAYBE(parent, maybeParent) {
+    parent->propagateTo(js, this->storage, kj::mv(maybeStorageEntry));
+  } else {
+    current(js).propagateTo(js, this->storage, kj::mv(maybeStorageEntry));
+  }
+}
+
+void AsyncContextFrame::propagateTo(
+    Lock& js,
+    Storage& other,
+    kj::Maybe<StorageEntry> maybeStorageEntry) {
+  storage.eraseAll([](const auto& entry) { return entry.key->isDead(); });
+  KJ_DASSERT(other.size() == 0);
+  for (auto& entry : storage) {
+    other.insert(entry.clone(js));
+  }
+
+  KJ_IF_MAYBE(entry, maybeStorageEntry) {
+    other.upsert(kj::mv(*entry), [](StorageEntry& existing, StorageEntry&& row) mutable {
+      existing.value = kj::mv(row.value);
+    });
+  }
+}
+
+kj::Maybe<AsyncContextFrame&> AsyncContextFrame::tryUnwrap(
+    Lock& js,
+    v8::Local<v8::Promise> promise) {
+  auto handle = v8::Private::ForApi(js.v8Isolate, v8StrIntern(js.v8Isolate, "asyncResource"));
```

Comment:
It looks like we will be re-allocating this handle (or at least looking up the interned version of it) for every .then(). Can we instead store a permanent handle to it somewhere?